### DISCUSSION TITLE
What is Metals doing under the hood in Emacs?

### DISCUSSION THREAD

**Discussion by kschwarz1116** [scalameta/metals#5951] [General]
I'm opening a discussion because I don't think I have a clear question here yet, but I'm trying to understand how Metals works in the context of Emacs.

The reason I'm asking is because my current workflow is:
* Start sbt and then `~compile` in process 1
* Start emacs in process 2 
* Make changes responding to the compile output in process 1

I've discovered recently a situation where sbt is showing my source as compiling successfully, but when I go into `lsp-metals-doctor-run` I see:
```
 📁   compiler error  14 Dec 22:08:16  scala.reflect.internal.Types$TypeError: illegal cyclic reference
                                       involving object implicits
```
`compiler error` links to a `.md` file that I can't act on. So it seems like Metals is compiling my code separately from _my_ sbt process somehow. I'm lost on how to investigate that though.

The installation instructions add in quite a few things that look like they _aren't_ Metals. I'd appreciate some thoughts as to why they aren't managed as `lsp-metals` dependencies, and also maybe some more description as to which parts of the experience provided are Metals, and which parts of the experience are provided by other packages.

**Comment by tgodzik** [comment]
I am not 100% sure about Emacs, but as for:
```
compiler error links to a .md file that I can't act on. So it seems like Metals is compiling my code separately from my sbt process somehow. I'm lost on how to investigate that though.
```
This is a presentation compiler error, which Metals uses underneath to provide completions, hovers and much more. This is part of the normal compiler and it is currently used only on the currently focused file. Since, the code might often break, some errors might be false positives, though ideally the compiler should never throw. If the code completions and the rest of the features are not working and you only see the errors popping up, this means there's something wrong. This could be classpath issues or compiler options causing some unexpected beahviour.

**Reply by kschwarz1116** [reply]
Interesting, is there a way to use metals to generate an sbt command that would execute the presentation compiler outside of metals? I'm trying to figure out if I can recreate the state to get the error I'm seeing outside of the editor.

**Reply by tgodzik** [reply]
It's not, the presentation compiler is not supposed to be for normal compilation, but the main reason here is just that we get up to a typer and there is some error handling to make trees at least partly correct in case of compilation errors.

This exception should be reproducible normally. If not it means we invoke compilation with some different setting compared to sbt.

**Reply by kschwarz1116** [reply]
So I've started to look into lsp-mode a bit and figure out what the mode-line is telling me. My (presentation compiler) errors seemed to be related to the fact that I was using a plugin that was doing code generation to drop scala files into my managed source. After I stopped using it, metals still picked up issues in those managed files that sbt was ignoring. Fortunately an sbt clean fixed this.

I'll try experimenting a bit to see if I can reproduce this with regularity.
